File: 03_Arrays/Easy/C++/04_remove_duplicates.cpp
// remove duplicates from the Sorted Array
// and return the number of unique elements

// make changes in the array itself

#include <bits/stdc++.h>
using namespace std;

int removeDuplcates_Brute(vector<int> &nums){
// brute force
  set<int> st;

  // store the unique elements in set
  for(int i=0; i<nums.size(); ++i){ // -> O(N) 
    st.insert(nums[i]); // -> O(LogN)
  }

  // till here TC -> O(NlogN)

  int index = 0;
  // O(N)
  for(auto it : st){
    nums[index] = it;
    ++index;
  }

  // total TC -> O(NlogN + N)
  // total SC -> O(N)
  return index;
}

int removeDuplcates_Optimal(vector<int> &nums){
  // optimal
  // -> 2 ptr approach
    // total TC -> O(N)
    // total SC -> O(1)

  
  int i = 0;
  for (int j = 1; j < nums.size(); ++j){
    // if j ptr encounters another unique element
    // store it
    // and increment i
    if (nums[j] != nums[i]){
      nums[i+1] = nums[j];
      ++i;
    }
  }

  // size of unique array would be 
  // index of last unique element + 1
  return i+1;

        
}

int main(int argc, char* argv[]){
  vector<int> v;

  int size;
  cin >> size;

  for(int i=0; i<size; ++i){
    int a;
    cin >> a;
    v.push_back(a);
  }

  cout << removeDuplcates_Brute(v) << "\n";
  cout << removeDuplcates_Optimal(v) << endl ;

return 0;
}

File: 03_Arrays/Easy/Python/04_remove_duplicates.py
# remove duplicates from the Sorted Array
# and return the number of unique elements

# make changes in the array itself

def remove_duplicates_brute(arr: list) -> int:
  st = set()

  # TC -> O(N)
  for i in range(len(arr)):
    st.add(arr[i])
  
  # as IN PYTHON: sets do not support indexing
  # so we'll convert the set into a list
  unique_list = sorted(st) # unorderd set -> sorted list
  # TC -> O(NlogN)

  # TC -> O(N)
  for j in range(len(unique_list)):
    arr[j] = unique_list[j]
  

  # Total TC -> O(NlogN + 2N)
  # Total SC -> O(N)
  return len(unique_list)


# optimal approach
    # using -> 2 ptr
def remove_duplicates_optimal(arr: list) -> int:

  i = 0
  for j in range(1, len(arr)):
    if (arr[j] != arr[i]):
      i += 1
      arr[i] = arr[j]
  
  # TC -> O(N)
  # SC -> O(1)
  return i+1


size = int(input("Enter size: "))
arr = list(map(int, input("Enter elements in array: ").strip().split()))

# newSize = remove_duplicates_brute(arr)
newSize = remove_duplicates_optimal(arr)

print(arr)
print(arr[:newSize])